src/Bicep.LangServer/Completions/BicepCompletionProvider.cs

```diff
@@ -291,31 +291,71 @@ private IEnumerable<CompletionItem> GetModulePathCompletions(SemanticModel model
-                .WithCommand(new Command { Name = EditorCommands.RequestCompletions })
-                .Build())
-                .ToList();
-            return fileItems.Concat(dirItems);
+                .Select(dir =>
+                    CreateModulePathCompletionBuilder(
+                        dir.Segments.Last(),
+                        // "./" will not be preserved when making relative Uris. We have to go and manually add it.
+                        (entered.StartsWith("./") ? "./" : "") + cwdUri.MakeRelativeUri(dir).ToString(),
+                        context.ReplacementRange,
+                        CompletionItemKind.Folder,
+                        CompletionPriority.Low)
+                    .WithCommand(new Command { Name = EditorCommands.RequestCompletions })
+                    .Build());
+
+            return bicepFileItems.Concat(armTemplateFileItems).Concat(dirItems);
+
+            // Local functions.
+            IEnumerable<CompletionItem> CreateFileCompletionItems(IEnumerable<Uri> fileUris, Uri cwdUri, Predicate<Uri> predicate, CompletionPriority priority) => fileUris
+                .Where(fileUri => fileUri != model.SourceFile.FileUri && predicate(fileUri))
+                .Select(fileUri =>
+                    CreateModulePathCompletionBuilder(
+                        fileUri.Segments.Last(),
+                        (entered.StartsWith("./") ? "./" : "") + cwdUri.MakeRelativeUri(fileUri).ToString(),
+                        context.ReplacementRange,
+                        CompletionItemKind.File,
+                        priority)
+                    .Build());
+
+            bool IsBicepFile(Uri fileUri)
+            {
+                if (PathHelper.HasBicepExtension(fileUri))
+                {
+                    return true;
+                }
+
+                return model.Compilation.SourceFileGrouping.SourceFiles.Any(sourceFile =>
+                    sourceFile is BicepFile &&
+                    sourceFile.FileUri.LocalPath.Equals(fileUri.LocalPath, PathHelper.PathComparison));
+            }
+
+            bool IsArmTemplateFileLike(Uri fileUri)
+            {
+                if (PathHelper.HasExtension(fileUri, LanguageConstants.ArmTemplateFileExtension))
+                {
+                    return true;
+                }
+
+                if (model.Compilation.SourceFileGrouping.SourceFiles.Any(sourceFile =>
+                        sourceFile is ArmTemplateFile &&
+                        sourceFile.FileUri.LocalPath.Equals(fileUri.LocalPath, PathHelper.PathComparison)))
+                {
+                    return true;
+                }
+
+                if ((PathHelper.HasExtension(fileUri, LanguageConstants.JsonFileExtension) ||
+                    PathHelper.HasExtension(fileUri, LanguageConstants.JsoncFileExtension)) &&
+                    this.FileResolver.TryRead(fileUri, out var fileContents, out var _, Encoding.UTF8, 2000, out var _))
```

Comment:
Any perf concerns with this approach? It seems fast enough to me with ~500 files on a new-ish MacBook with an SSD, but could slow down with larger quantities of files or HDDs / other unusual disk configurations.

I get the need for it (as you don't want param files showing up), so not sure if we really have another option!

Reply:
Yeah this can definitely be slow with HDDs if there are a hundreds of files in the directory. Although I cannot think of a way to avoid reading the files (as you mentioned we don't want random JSON files showing up in the completion list), we are only checking the first 2000 characters of each JSON file, so the delay of loading completions is linear to the number of files and is not affected by file sizes.

Reply:
I suspect we can go lower than 2000. The schema property is usually at the beginning and isn't that long. If the user decides to put it at the end of the long template, I think it's ok if the file doesn't show up in completions. For those cases, the user shouldn't be blocked because they can still write in a valid path.

I think there are a couple of options to preserve the experience if perf ends up being slow for some users:
1. We can put in a hard time limiter in the logic. It could be as simple as starting a stopwatch. If elapsed time exceeds some threshold, we stop reading more file and don't return any of the JSON completions.
2. We are already receiving notifications about all file events in a workspace (and discarding most of them). We could actually keep track of JSON files specifically and load them only when they are modified to determine if they are a template file. Then the completion logic would just have to query the cache.